### ISSUE TITLE
Show package name in hover

### ISSUE THREAD

**Issue by Thibaulltt** [astral-sh/ty#2967]
In my codebase, I have a `<my library>.Tensor` type which I use extensively in typing hints, which is defined as such:

```python
Tensor = Union[numpy.ndarray, torch.Tensor]
```

... since I have a set of operations that should work regardless of the particular API a user is using. When I look at the code either using this API or within my library, I often use ty to infer if the type of the variable I'm looking at is right. This leads to:

```python
τ = torch.rand((10,10))          # 1
ν = numpy.random.random((10,10)) # 2
μ = my_library.foo(τ)            # 3
```

Hovering over the variables in the three lines above gives the following type hints/definitions upon `textDocument/hover`:

- `line# 1`: `Tensor`
- `line# 2`: `ndarray`
- `line# 3`: `Tensor`

Would it be possible to show the "fully qualified name" of the type instead of its class/instance name ? This would mean showing:

- `line# 1`: `torch.Tensor`
- `line# 2`: `numpy.ndarray`
- `line# 3`: `<my library>.Tensor`, or `numpy.ndarray`/`torch.Tensor` if static analysis allows it :)

**Comment by Thibaulltt** [comment]
This would also solve an issue I have while using `ty` to explore a new codebase. Whenever I open a new project and look at existing code using a library I'm not familiar with, I often hover over a symbol (class names/type hints) and trigger the `textDocument/hover` method in order to see where or in what module is this symbol defined.

For example, take a look at [`xyz2tex/svg2tikz@master:svg2tikz/tikz_export.py:L520`](https://github.com/xyz2tex/svg2tikz/blob/72efc44f5deea9ed329d3378f8fa6163dda3b5cd/svg2tikz/tikz_export.py#L520) :

```python
class TikZPathExporter(inkex.Effect, inkex.EffectExtension):
```

Hovering over the `inkex.Effect` symbol returns the following: `<class 'Effect'>`. This should instead return `<class 'inkex.deprecated.deprecatedeffect.Effect'` if we return the fully-qualified name.

**Comment by MichaReiser** [comment]
We do have the necessary information to render the full qualified name. We mainly don't use the full qualified name because it's long :) But I can understand that this can make it harder to determine where the type is coming from. Does ty's behavior differ here from the tool you used before? 

We do have an issue to make it easier to jump to the hovered type, see https://github.com/astral-sh/ty/issues/1575. THe idea is that you could click the type to go to its definition. But I'm not sure if this would help in your case.

**Comment by Thibaulltt** [comment]
Thanks for the quick response! It does differ from the built-in documentation and LSP built-in PyCharm, where I spent a lot of time browsing Python codebases at work. See the following example:

<img width="426" height="160" alt="Image showing a snippet of Python code. The cursor is positioned over a symbol named 'SurfaceMesh' and a tooltip showing its documentation is opened below." src="https://github.com/user-attachments/assets/c951f96c-b550-4a5f-ab56-2aed669c94b9" />

And since I edit more and more in `neovim` using `ty` as the LSP, it's a quality of life that's sorely missed :) You could maybe do something similar to what they do here, by returning a text akin to:

```
<class 'SurfaceMesh'>
from 'polyscope.SurfaceMesh'
```

While #1575 looks nice, since I mostly use `neovim` as stated above I already have a 'go-to-definition' key mapping, this issue mostly relates to being able to quickly gather where a symbol is coming from without leaving my file (and mental) context.

**Comment by MichaReiser** [comment]
That's helpful. Unlike Pycharm, We're , unfortunately, limited by what the [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.18/specification/#textDocument_hover) supports and hover only supports text (markdown for some editors) content. But yes, we could try to include the full qualified name somewhere. It gets more tricky for Unions and such that might include multiple types from different packages.

**Comment by Thibaulltt** [comment]
Why not put it as another `MarkedString` element, before/after the description ? Quickly looking at the [`neovim` implementation of their `textDocument/hover`](https://github.com/neovim/neovim/blob/9fa7c0a04638ee3d5fefbdcec347b1c07a5ce9ce/runtime/lua/vim/lsp/buf.lua#L53-L124), they take all `MarkedString` in the response and display them one on top of the other, leading to the following stacked tooltip:

<img width="719" height="147" alt="Image showing a snippet of Python text. The cursor is placed on the text 'inkex.Effect' and a tooltip is opened below, showing the name of the class and a quick description." src="https://github.com/user-attachments/assets/00b23f3c-cc9c-4e42-ae58-b83af5c9d1a6" />

This would only need to be another `MarkedString` at the end of the response you send, wouldn't it?

**Comment by MichaReiser** [comment]
This could work. I'd have to look at how other LSP render the message

**Comment by AlexWaygood** [comment]
In lots of our diagnostics, we show the full name if we can see from the context that the unqualified name would be ambiguous. E.g. here we don't display the confusing type `bool[bool]`, we show the type `library.bool[builtins.bool]`: https://play.ty.dev/cf535166-90e6-403b-8d1f-7ce917edcb1d. Could we do something similar for our on-hover display? In the situation in the OP here, it seems like we could detect that there are two different `Tensor` types mentioned on the same line and use that to inform whether or not we show the full name or not. But maybe that heuristic wouldn't be effective in lots of other situations.

**Comment by Thibaulltt** [comment]
I deliberately made the example contrived, but my setup currently defines the `Tensor` type in the `<my library>.utils._typing` submodule, and it's used in `<my library>.{foo, bar, baz}.*` submodules through direct imports :)

**Comment by HernandoR** [comment]
Hi, thanks for all the work on ty — and sorry to add to an already-long thread.

I wanted to mention a related angle in case it's useful: my use case is functions rather than type names. When I hover a function call I get the signature and the docstring, but not *where the function comes from*.

That matters to me for two reasons. 

The first is ambiguity: my own project often has functions and classes named the same as, or very close to, ones in a third-party library, and today the only way to resolve which is which is to scroll back to the imports at the top of the file — exactly the context switch I'd like the hover to save me. 

The second is that the code I'm reading is frequently not code I wrote. When I come across an unfamiliar call, knowing which package it originates from is a large part of how I judge it: whether it's a well-maintained dependency or something deprecated, whether it's a wrapper of our own that might have known rough edges, and generally whether it's worth being suspicious of. The signature alone doesn't carry any of that.

## Example

My project has `mypkg.data.registry:load_dataset`, with a signature very similar to `datasets.load_dataset`. Either one may be the one in scope, depending on the file:

```python
# near the top of a 400-line module, long since scrolled off screen
from datasets import load_dataset

...

ds = load_dataset(cfg.name)   # <- hovering here
```

The hover currently shows:

> ```python
> def load_dataset(path: str, name: str | None = None, ...) -> DatasetDict
> ```
> Load a dataset from the Hugging Face Hub, or a local dataset.

That's accurate, but it doesn't answer the question I actually have — *which* `load_dataset` is this? The signature and docstring of my own version read similarly enough that I still have to jump to the top of the file to be sure. One extra line would settle it without leaving my current position:

> ```python
> def load_dataset(path: str, name: str | None = None, ...) -> DatasetDict
> ```
> `datasets.load:load_dataset`
>
> Load a dataset from the Hugging Face Hub, or a local dataset.

The same applies to methods — hovering `ds.map(...)`:

> ```python
> def map(self, function: Callable | None = None, ...) -> Dataset
> ```
> `datasets.arrow_dataset:Dataset.map`

## On notation

Would you consider the entry-point specifier form (`module:qualname`) for that line?

```
mypkg.subpackage:function
mypkg.subpackage:Class.function
```

My thinking is that a uniform `.` can't show where the importable module path ends — in `datasets.arrow_dataset.Dataset.map` there's no way to tell which segments are submodules and which are attributes — and that boundary is exactly what I need in order to write or verify the import. The `module:qualname` form makes it explicit, and since it already has a packaging spec behind it, it shouldn't feel foreign to Python users.

Two small notes:

- The line probably wants to be plain markdown text rather than part of a ```python block, since `:` isn't valid Python there (cf. #2498).
- Keeping it on its own line rather than inline in the signature also leaves room in case the fully qualified name too long

Either way, thanks for considering it!
